light/client.go

```diff
@@ -940,52 +937,92 @@ func (c *Client) backwards(
 	return nil
 }
 
-// NOTE: requires a providerMutex locked.
-func (c *Client) removeWitness(idx int) {
-	switch len(c.witnesses) {
-	case 0:
-		panic(fmt.Sprintf("wanted to remove %d element from empty witnesses slice", idx))
-	case 1:
-		c.witnesses = make([]provider.Provider, 0)
+// lightBlockFromPrimary retrieves the lightBlock from the primary provider
+// at the specified height. This method also handles provider behavior as follows:
+//
+// 1. If the provider does not respond or does not have the block, it tries again
+//    with a different provider
+// 2. If all providers return the same error, the light client forwards the error to
+//    where the initial request came from
+// 3. If the provider provides an invalid light block, is deemed unreliable or returns
+//    any other error, the primary is permanently dropped and is replaced by a witness.
+func (c *Client) lightBlockFromPrimary(ctx context.Context, height int64) (*types.LightBlock, error) {
+	c.providerMutex.Lock()
+	l, err := c.primary.LightBlock(ctx, height)
+	c.providerMutex.Unlock()
+
+	switch err {
+	case nil:
+		// Everything went smoothly. We reset the lightBlockRequests and return the light block
+		c.lightBlockRequests = 0
+		return l, nil
+
+	case provider.ErrNoResponse:
+	case provider.ErrLightBlockNotFound:
+		c.lightBlockRequests++
+		// if the light client has requested the same block to all providers and they have given one of
+		// the above two errors, then we can assume that the light block is unreachable at the time and
+		// we return the error
+		if c.lightBlockRequests > uint16(len(c.witnesses)) {
+			return nil, err
+		}
+		// else we cycle through the next witness and try again
+		c.swapPrimaryProvider()
+
 	default:
-		c.witnesses[idx] = c.witnesses[len(c.witnesses)-1]
-		c.witnesses = c.witnesses[:len(c.witnesses)-1]
+		// The light client has most likely received either provider.ErrUnreliableProvider or provider.ErrBadLightBlock
+		// These errors mean that the light client should drop the primary and try with another provider instead
+		c.logger.Debug("Error on light block request from primary", "error", err, "primary", c.primary)
+		if removeErr := c.removePrimaryProvider(); removeErr != nil {
+			return nil, fmt.Errorf("%v. Tried to remove primary but: %w", err.Error(), removeErr)
+		}
 	}
+
+	return c.lightBlockFromPrimary(ctx, height)
 }
 
-// replaceProvider takes the first alternative provider and promotes it as the
-// primary provider.
-func (c *Client) replacePrimaryProvider() error {
+// NOTE: requires a providerMutexLocked
+func (c *Client) removeWitness(idx int) error {
+	if len(c.witnesses) < 2 {
+		return ErrNoWitnesses
```

Comment:
why c.witnesses == 1 constitutes as no witnesses?

Reply:
because we are just about to remove that witness. Thus at the end of this function we will have zero witnesses